mobile/src/foss/java/org/openhab/habdroid/core/MessageListenerService.kt

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.openhab.habdroid.core
+
+import android.app.NotificationManager
+import android.content.Context
+import android.content.Context.NOTIFICATION_SERVICE
+import android.util.Log
+import androidx.core.content.edit
+import kotlinx.coroutines.runBlocking
+import org.json.JSONArray
+import org.json.JSONException
+import org.openhab.habdroid.core.connection.ConnectionFactory
+import org.openhab.habdroid.model.toCloudNotification
+import org.openhab.habdroid.util.HttpClient
+import org.openhab.habdroid.util.NotificationUtils.HAS_GROUPING_SUPPORT
+import org.openhab.habdroid.util.NotificationUtils.SUMMARY_NOTIFICATION_ID
+import org.openhab.habdroid.util.NotificationUtils.createChannelForSeverity
+import org.openhab.habdroid.util.NotificationUtils.getChannelId
+import org.openhab.habdroid.util.NotificationUtils.getGcmNotificationCount
+import org.openhab.habdroid.util.NotificationUtils.makeNotification
+import org.openhab.habdroid.util.NotificationUtils.makeSummaryNotification
+import org.openhab.habdroid.util.PrefKeys
+import org.openhab.habdroid.util.getPrefs
+import org.openhab.habdroid.util.map
+
+class MessageListenerService {
+    companion object {
+        private val TAG = MessageListenerService::class.java.simpleName
+
+        suspend fun checkForMessages(context: Context) {
+            ConnectionFactory.waitForInitialization()
+            val connection = ConnectionFactory.cloudConnectionOrNull
+            if (connection == null) {
+                Log.d(TAG, "Got no connection")
+                return
+            }
+            val url = "api/v1/notifications?limit=20"
+            var messages = try {
+                val response = connection.httpClient.get(url).asText().response
+                Log.d(TAG, "Notifications request success")
+                JSONArray(response).map { obj -> obj.toCloudNotification() }
+            } catch (e: JSONException) {
+                Log.d(TAG, "Notification response could not be parsed", e)
+                null
+            } catch (e: HttpClient.HttpException) {
+                Log.e(TAG, "Notifications request failure", e)
+                null
+            } ?: return
+
+            val prefs = context.getPrefs()
+
+            val lastShownMessage = prefs.getString(PrefKeys.FOSS_LAST_SHOWN_MESSAGE, "-1")!!.toBigInteger(16)
```

Comment:
I don't think messing around with `BigInteger` is actually needed. I think it would be simpler and more logical to find the index of the last seen notification ID in the received list (using indexOf) and only use notifications up to that index using `subList()`.
I'd also suggest keeping the pref key name constant internal to this class, as it's not used anywhere else.
Basically this:
```
val lastSeenMessageId = prefs.getString(LAST_SEEN_MESSAGE_KEY, null)
prefs.edit {
    val newestSeenId = messages.firstOrNull()?.id ?: lastSeenMessageId
    putString(LAST_SEEN_MESSAGE_KEY, newestSeenId)
}
if (lastSeenMessageId == null) {
    // Never checked for notifications before
    return
}
val lastSeenIndex = messages.indexOf(lastSeenMessageId)
val newMessages = if(lastSeenIndex >= 0) messages.subList(0, lastSeenIndex) else messages
if (newMessages.isEmpty()) {
    // out
}
// normal handling
```